File: PrismaApi/PrismaApi.Domain/Dtos/UserDtos.cs
using System;
using System.Collections.Generic;
using System.Text.Json.Serialization;

namespace PrismaApi.Domain.Dtos;

public class UserDto
{
    [JsonPropertyName("user_id")]
    public required string Id { get; set; }
    [JsonPropertyName("name")]
    public string Name { get; set; } = string.Empty;
}

public class UserIncomingDto : UserDto
{
    [JsonPropertyName("project_roles")]
    public List<ProjectRoleIncomingDto> ProjectRoles { get; set; } = new();
}

public class UserOutgoingDto : UserDto
{

}

File: src/dtos/user_dtos.py
from pydantic import BaseModel, Field
from typing import Optional, Annotated
from src.dtos.project_roles_dtos import (
    ProjectRoleIncomingDto,
    ProjectRoleOutgoingDto,
    ProjectRoleMapper,
)
from src.models.user import User
from src.constants import DatabaseConstants


class UserDto(BaseModel):

    name: Annotated[str, Field(max_length=DatabaseConstants.MAX_SHORT_STRING_LENGTH.value)]
    azure_id: Annotated[str, Field(max_length=DatabaseConstants.MAX_SHORT_STRING_LENGTH.value)]


class UserIncomingDto(UserDto):
    user_id: Optional[int]
    project_roles: list[ProjectRoleIncomingDto] = []


class UserOutgoingDto(UserDto):
    user_id: int
    project_roles: list[ProjectRoleOutgoingDto] = []


class UserMapper:
    @staticmethod
    def to_outgoing_dto(entity: User) -> UserOutgoingDto:
        return UserOutgoingDto(
            user_id=entity.id,
            name=entity.name,
            azure_id=entity.azure_id,
            project_roles=ProjectRoleMapper.to_outgoing_dtos(entity.project_role),
        )

    @staticmethod
    def to_entity(dto: UserIncomingDto) -> User:
        return User(id=dto.user_id, name=dto.name, azure_id=dto.azure_id, project_role=[])

    @staticmethod
    def to_entities(dtos: list[UserIncomingDto]) -> list[User]:
        return [UserMapper.to_entity(dto) for dto in dtos]

    @staticmethod
    def to_outgoing_dtos(entities: list[User]) -> list[UserOutgoingDto]:
        return [UserMapper.to_outgoing_dto(entity) for entity in entities]
